### ISSUE TITLE
DXIL validation error but no hint as to why

### ISSUE THREAD

**Issue by siliconvoodoo** [microsoft/DirectXShaderCompiler#2457]
I tried that into whatever version is currently hosted on Tim Jones's playground.

```
struct PSInput
{
	float4 color : COLOR;
};

float4 other(out const int);

float4 other(out const int)
{
}

float4 PSMain(PSInput input) : SV_TARGET
{
    int i;
    other(i);
    return input.color;
}
```

And the result is:

> warning: DXIL.dll not found.  Resulting DXIL will not be signed for use in release environments.
> 
> error: validation errors
> at 0x1d416afa230 inside block #0 of function PSMain Instructions must be of an allowed type

Which puzles me as to what is wrong in the original code, what should I focus on to fix ?

**Comment by tex3d** [comment]
There is a bug causing bad code gen in this case, where you have a function with a non-void return type, but do not return anything from the function.  We should report an error earlier in this case.  

Also, note that this function has an `out const int` argument.  `out` means you must assign to this output parameter in the function, but `const` says you cannot assign to this argument, so this is a contradiction, which we should also have an error message for.

**Comment by tex3d** [comment]
Fxc behavior:
`float4 other(out const int);` - does not support unnamed parameters, so:
`error X3000: syntax error: unexpected token ')'`

`float4 other(out const int i);` - if you add a name:
`error X3046: 'i': output parameters cannot be declared 'const'`

`float4 other() { }`:
`error X3507: 'other': Not all control paths return a value`

**Comment by siliconvoodoo** [comment]
thanks. yes as you imagine, this snippet is a deliberate "attack" on the language to test how it behaves. what I meant by "puzzle" is usually the build tools tell us what we did wrong, ideally at the exact syntactic position it started to hit the fan. As you detected, here you lack 2 messages; I guess it's relatively easy to improve on your front end.

**Comment by tex3d** [comment]
Since we have a separate bug on the crash from no return, I'll call this enhancement to track adding error detection/handling for `output parameters cannot be declared 'const'`.

**Comment by pow2clk** [comment]
The current behavior of the driver is to accept the `const out` qualifiers without complaint. However, if you try to alter the parameter, it complains that it is const. So really the only flaw is if you declare such a parameter and then don't assign it.

**Comment by pow2clk** [comment]
In triage, we determined that the only remaining issue is not worth fixing.
